packages/flame/lib/src/shader_pipeline/shader_pipeline.dart

```diff
@@ -0,0 +1,362 @@
+import 'dart:ui' as ui show Image;
+import 'dart:ui' hide Image;
+
+import 'package:flame/components.dart';
+import 'package:flutter/foundation.dart';
+
+
+
+abstract class SPipelineStep extends PositionComponent {
+  SPipelineStep({
+    
+    required this.samplingPasses,
+    double? pixelRatio,
+    super.position,
+    super.size,
+    super.scale,
+    super.angle,
+    double? nativeAngle,
+    super.anchor,
+    super.children,
+    super.priority,
+    super.key,
+  }) :  pixelRatio = pixelRatio ??
+            PlatformDispatcher.instance.views.first.devicePixelRatio,
+            super(
+          nativeAngle: nativeAngle ?? 0,
+        );
+
+final int samplingPasses;
+
+
+
+
+  late final _layer = _SPipelineStepLayer(
+    renderTree: super.renderTree,
+    canvasFactory: _canvasFactory,
+    samplingPasses: samplingPasses,
+    postProcess: postProcess,
+  );
+
+
+  
+  final double pixelRatio;
+
+  Canvas _canvasFactory(PictureRecorder recorder, int passIndex) {
+    return SamplingCanvas(
+      step: this,
+      passIndex: passIndex,
+      actualCanvas: Canvas(recorder),
+    );
+  }
+
+  @override
+  void renderTree(Canvas canvas) {
+    decorator.applyChain(
+      (canvas) => _layer.render(canvas, size, pixelRatio),
+      canvas,
+    );
+  }
+
+  void postProcess(List<ui.Image> samples, Size size, Canvas canvas);
+}
+
+class _SPipelineStepLayer {
+  final Canvas Function(PictureRecorder, int) canvasFactory;
+  final void Function(Canvas) renderTree;
+  final void Function(List<ui.Image>, Size, Canvas) postProcess;
+  final int samplingPasses;
+
+  _SPipelineStepLayer({
+    required this.renderTree,
+    required this.canvasFactory,
+    required this.samplingPasses,
+    required this.postProcess,
+  });
+
+  ui.Image _renderPass(Vector2 size, int pass, double pixelRatio) {
+    final recorder = PictureRecorder();
+
+    final innerCanvas = canvasFactory(recorder, pass);
+    renderTree(innerCanvas);
+    final picture = recorder.endRecording();
+
+    return picture.toImageSync(
+      (pixelRatio * size.x).ceil(),
+      (pixelRatio * size.y).ceil(),
+    );
+  }
+
+  void render(Canvas canvas, Vector2 size, double pixelRatio) {
+    final results = List<ui.Image>.generate(
+      samplingPasses,
+      (i) {
+        return _renderPass(size, i, pixelRatio);
+      },
+    );
+
+    canvas.save();
+    postProcess(results, size.toSize(), canvas);
+    canvas.restore();
+  }
+}
+
+class SamplingCanvas<S extends SPipelineStep> implements Canvas {
```

Comment:
could consider some sort of "context" api - but we can have this version as is as a first one